daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Normalization.scala

```diff
@@ -0,0 +1,153 @@
+    * - field names are dropped from Records
+    * - values are normalized recursively
+    * - all values embedded in transaction nodes are normalized
+    * - version-specific normalization is applied to the 'byKey' fields of 'NodeFetch' and 'NodeExercises'
+    *
+    * We do not normalize the node-ids in the transaction here, but rather assume that
+    * aspect of normalization has already been performed (by the engine, or by
+    * deserialization).
+    *
+    * Eventually we would like that all aspects of normalization are achieved directly by
+    * the transaction which is constructed by the engine. When this is done, we will no
+    * longer need this separate normalization pass.
+    */
+
+  private type Val = V[Cid]
+  private type KWM = KeyWithMaintainers[Val]
+  private type Node = GenNode[Nid, Cid]
+  private type VTX = VersionedTransaction[Nid, Cid]
+
+  def normalizeTx(vtx: VTX): VTX = {
+    vtx match {
+      case VersionedTransaction(_, nodes, roots) =>
+        // TODO: Normalized version calc should be shared with code in asVersionedTransaction
+        val version = roots.iterator.foldLeft(TransactionVersion.minVersion) { (acc, nodeId) =>
+          import scala.Ordering.Implicits.infixOrderingOps
+          nodes(nodeId).optVersion match {
+            case Some(version) => acc max version
+            case None => acc max TransactionVersion.minExceptions
+          }
+        }
+        VersionedTransaction(
+          version,
+          nodes.map { case (k, v) =>
+            (k, normNode(v))
+          },
+          vtx.roots,
+        )
+    }
+  }
+
+  private def normNode(
+      node: Node
+  ): Node = {
+    import scala.Ordering.Implicits.infixOrderingOps
+    node match {
+
+      case old: NodeCreate[_] =>
+        old
+          .copy(arg = normValue(old.arg))
+          .copy(key = old.key.map(normKWM))
+
+      case old: NodeFetch[_] =>
+        (if (old.version >= TransactionVersion.minByKey) {
+           old
+         } else {
+           old.copy(byKey = false)
+         })
+          .copy(
+            key = old.key.map(normKWM)
+          )
+
+      case old: NodeExercises[_, _] =>
+        (if (old.version >= TransactionVersion.minByKey) {
+           old
+         } else {
+           old.copy(byKey = false)
+         })
+          .copy(
+            chosenValue = normValue(old.chosenValue),
+            exerciseResult = old.exerciseResult.map(normValue),
+            key = old.key.map(normKWM),
+          )
+
+      case old: NodeLookupByKey[_] =>
+        old.copy(
+          key = normKWM(old.key)
+        )
+
+      case old: NodeRollback[_] => old
+
+    }
+  }
+
+  // It's ok to code this in stack unaware fashion, as values  have a depth limit of 100
+  private def normValue(x: Val): Val = {
+    x match {
+
+      // recursive cases (with normalization)
+      case V.ValueRecord(_, fields) =>
+        V.ValueRecord(
+          tycon = None, //norm
```

Comment:
That’s not right. We should only drop the type constructor for LF >= 1.12.

Reply:
done
(by switching to existing code for doing value normalization)